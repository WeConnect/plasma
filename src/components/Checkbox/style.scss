@import '../../styles/base';

.wrapperBase {
  position: relative;
  display: inline-flex;
  flex-direction: row;
  justify-content: flex-start;
  align-items: center;
  cursor: pointer;
}

.wrapper {
  composes: wrapperBase;
  &:hover {
    .checkbox {
      border: 1px solid $cGray200;
    }
  }
}

.wrapperDisabled {
  composes: wrapperBase;
  cursor: not-allowed;
  .original, .indeterminate {
    cursor: not-allowed;
  }
}


.original {
  position: absolute;
  opacity: 0;
  z-index: 10;
  cursor: pointer;
  ~ .checkbox::before {
    width: 12px;
    height: 8px;
    content: '';
  }

  &:checked ~ .checkbox {
    &::before {
      height: 20px;
      // content: url('data:image/svg+xml;utf8,<svg width="12px" height="10px" viewBox="2 5 12 10" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"><defs></defs><polygon id="Rectangle-34" stroke="none" fill="#282829" fill-rule="evenodd" points="6 15 2 10.8333333 3.6 9.16666667 6 11.6666667 12.4 5 14 6.66666667"></polygon></svg>');
      content: url('data:image/svg+xml;utf8,<svg width="12px" height="10px" viewBox="0 0 12 10" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"><!-- Generator: Sketch 43.1 (39012) - http://www.bohemiancoding.com/sketch --><title>checked</title><desc>Created with Sketch.</desc><defs></defs><g id="Symbols" stroke="none" stroke-width="1" fill="none" fill-rule="evenodd"><g id="checkbox-/-selected" transform="translate(-2.000000, -8.000000)" fill="#252729"><polygon id="checked" points="6 18 2 13.8333333 3.6 12.1666667 6 14.6666667 12.4 8 14 9.66666667"></polygon></g></g></svg>');
    }
  }
}

.indeterminate {
  position: absolute;
  opacity: 0;
  z-index: 10;
  cursor: pointer;

  ~ .checkbox::before {
    content: '';
    height: 2px;
  }

  &:checked ~ .checkbox {
    &::before {
      display: block;
      width: 8px;
      height: 2px;
      background: $cWeWorkBlack;
      position: relative;
      top: 0;
    }
  }
}

.checkbox {
  width: 16px;
  height: 16px;
  background: $cWhite;
  border: 1px solid $cBorder;
  border-radius: 3px;
  background: $cWhite;
  transition: border-color 0.1s ease-out;
  box-sizing: border-box;
  display: flex;
  justify-content: center;
  align-items: center;
  &::before {
    display: block;
    content: '';
    background: none;
  }
}

.textBase {
  margin-left: 10px;
  @extend %type--body;
}

.text {
  composes: textBase;
  cursor: pointer;
}

.textBold {
  font-weight: bold;
}

.textDisabled {
  composes: textBase;
  color: $cGray200;
}

.description {
  display: inline-block;
  @extend %fontFamily--body;
  font-size: 11px;
  color: $cGray200;
  line-height: 20px;
  margin-left: 26px;
}

