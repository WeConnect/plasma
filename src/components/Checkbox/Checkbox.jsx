import { uniqueId } from 'lodash';
import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import style from './style.scss';
import {
  getDataAttrs,
  getDataProps }
from '../../dataUtils';

class Checkbox extends React.Component {

  render() {
    const {
      isBold,
      checked,
      indeterminate,
      name,
      text,
      onChange,
      disabled,
      value,
      data,
    } = this.props;
    const id = uniqueId('id');
    const inputClassName = indeterminate ? style.indeterminate : style.original;
    const wrapperClassName = cx(style.wrapper, {
      [style.wrapperDisabled]: disabled,
    });
    const textClassName = cx(style.text, {
      [style.textBold]: isBold,
      [style.textDisabled]: disabled,
    });

    return (
      <div
        {...getDataAttrs(data)}
      >
        <div>
          <label
            htmlFor={id}
            className={wrapperClassName}
          >
            <input
              disabled={disabled}
              checked={checked || value}
              className={inputClassName}
              type="checkbox"
              id={id}
              name={name}
              onChange={onChange}
            />
            <div className={style.checkbox} />
            <div className={textClassName}>
              {text}
            </div>
          </label>
        </div>
        { this.props.description &&
          <span className={style.description}>
            { this.props.description }
          </span>
        }
      </div>
    );
  }
}

Checkbox.defaultProps = {
  checked: false,
  value: false,
  indeterminate: false,
  text: 'Option',
  name: 'checkbox',
};

Checkbox.propTypes = {
  checked: PropTypes.bool,
  value: PropTypes.bool,
  indeterminate: PropTypes.bool,
  text: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.node,
    PropTypes.element,
  ]).isRequired,
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func,
  disabled: PropTypes.bool,
  description: PropTypes.string,
  ...getDataProps(),
};

Checkbox.displayName = 'Plasma@Checkbox';

export default Checkbox;
