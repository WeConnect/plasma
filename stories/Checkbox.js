import React from 'react';
import { storiesOf } from '@storybook/react';

import StoryState from '../StoryState';
import Checkbox from '../src/components/Checkbox/Checkbox';
import meetupLogo from '../src/icons/meetupLogo';

storiesOf('Checkbox', module)
  .add('default', () => (
    <Checkbox />
  ))
  .add('with description', () => (
    <Checkbox description={'Hint: this is some text.'} />
  ))
  .add('checked', () => (
    <Checkbox checked />
  ))
  .add('unchecked', () => (
    <Checkbox checked={false} />
  ))
  .add('indeterminate unchecked', () => (
    <Checkbox indeterminate />
  ))
  .add('indeterminate checked', () => (
    <Checkbox indeterminate checked />
  ))
  .add('icon', () => (
    <StoryState render={(state, setState) => (
      <Checkbox
        icon={meetupLogo}
        name="optionOne"
        text="Meetup"
        onChange={e => setState({selected: e.target.name})}
        checked={state.selected === 'optionOne'}
      />
    )}
    />
  ))
  .add('fancy', () => (
    <StoryState render={(state, setState) => (
      <Checkbox
        fancy
        name="optionOne"
        onChange={e => setState({selected: e.target.name})}
        checked={state.selected === 'optionOne'}
      />
    )}
    />
  ))
  .add('multiple fancy with optional icon', () => (
    <div>
      <StoryState render={(state, setState) => (
        <Checkbox
          fancy
          name="dog"
          onChange={e => setState({selected: e.target.name})}
          text="Dog"
          checked={state.selected === 'dog'}
        />
      )}
      />
      <StoryState render={(state, setState) => (
        <Checkbox
          fancy
          name="cat"
          onChange={e => setState({selected: e.target.name})}
          text="Cat"
          checked={state.selected === 'cat'}
        />
      )}
      />
      <StoryState render={(state, setState) => (
        <Checkbox
          fancy
          name="snake"
          onChange={e => setState({selected: e.target.name})}
          text="Snake"
          checked={state.selected === 'snake'}
        />
      )}
      />
      <StoryState render={(state, setState) => (
        <Checkbox
          fancy
          icon={meetupLogo}
          name="optionOne"
          text="Meetup"
          onChange={e => setState({selected: e.target.name})}
          checked={state.selected === 'optionOne'}
        />
      )}
      />
    </div>
  ));
;
