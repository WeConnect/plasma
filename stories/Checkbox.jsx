import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';

import StoryState from '../StoryState';
import Checkbox from '../src/components/Checkbox/Checkbox';

storiesOf('Checkbox', module)
  .add('default', () => (
    <StoryState
      render={(state, setState) => (
        <Checkbox
          onChange={() => {
            setState({ checked: !state.checked });
            action('change');
          }}
        >
          Option
        </Checkbox>
      )}
    />
  ))
  .add('checked', () => <Checkbox checked>Option</Checkbox>)
  .add('unchecked', () => <Checkbox checked={false}>Option</Checkbox>)
  .add('disabled', () => <Checkbox disabled>Option</Checkbox>)
  .add('disabled checked', () => (
    <Checkbox disabled checked>
      Option
    </Checkbox>
  ))
  .add('error', () => <Checkbox error>Option</Checkbox>)
  .add('error checked', () => (
    <Checkbox error checked>
      Option
    </Checkbox>
  ));
