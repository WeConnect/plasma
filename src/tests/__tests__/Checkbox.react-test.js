/* eslint-disable */

import React from 'react';
import Checkbox from '../../components/Checkbox/Checkbox';
import renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';

describe('Checkbox Component', () => {
  test('render', () => {
    const component = renderer.create(
      <Checkbox />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
    tree.props.isChecked = true;
    expect(tree).toMatchSnapshot();
  });
});