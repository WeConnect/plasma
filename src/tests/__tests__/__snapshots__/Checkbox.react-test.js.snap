// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox Component render 1`] = `
<div>
  <label
    className="wrapper"
    htmlFor="id1"
  >
    <input
      checked={false}
      className="original"
      disabled={undefined}
      id="id1"
      name="checkbox"
      onChange={undefined}
      type="checkbox"
    />
    <div
      className="checkbox"
    />
    <div
      className="text"
    >
      Option
    </div>
  </label>
</div>
`;

exports[`Checkbox Component render 2`] = `
<div
  isChecked={true}
>
  <label
    className="wrapper"
    htmlFor="id1"
  >
    <input
      checked={false}
      className="original"
      disabled={undefined}
      id="id1"
      name="checkbox"
      onChange={undefined}
      type="checkbox"
    />
    <div
      className="checkbox"
    />
    <div
      className="text"
    >
      Option
    </div>
  </label>
</div>
`;
